cppcache/integration-test-2/TcrMessageTest.cpp

```diff
@@ -0,0 +1,277 @@
+
+    auto input = cache.createDataInput(buffer, sizeof(buffer));
+
+    message.readCallbackObjectPart(input, false);
+
+    const auto&& dataSerializablePrimitive = std::dynamic_pointer_cast<DataSerializablePrimitive>(message.getCallbackArgumentRef());
+    ASSERT_THAT(dataSerializablePrimitive.get(), ::testing::NotNull());
+    auto output = cache.createDataOutput();
+    dataSerializablePrimitive->toData(output);
+    EXPECT_EQ(output.getBufferLength(), 4);
+
+    uint8_t b[] = {3, 'a','b','c'};
+    EXPECT_THAT(b, ::testing::ElementsAreArray(output.getBuffer(), output.getBufferLength()));
+}
+
+TEST(TcrMessage, readSecureObjectPart1) {
+    auto cache = CacheFactory().create();
+    TcrMessageDummy message;
+    uint8_t buffer[] = {0, 0, 0, 1, //Size
+                        1,         //isObj
+                        static_cast<uint8_t>(DSCode::CacheableBytes),
+                        3,
+                        'a', 'b', 'c' };
+
+    auto input = cache.createDataInput(buffer, sizeof(buffer));
+
+    message.readSecureObjectPart(input);
+
+    const auto&& dataSerializablePrimitive = std::dynamic_pointer_cast<DataSerializablePrimitive>(message.getValueRef());
+    ASSERT_THAT(dataSerializablePrimitive.get(), ::testing::NotNull());
+    auto output = cache.createDataOutput();
+    dataSerializablePrimitive->toData(output);
+    EXPECT_EQ(output.getBufferLength(), 4);
+
+    uint8_t b[] = {3, 'a','b','c'};
+    EXPECT_THAT(b, ::testing::ElementsAreArray(output.getBuffer(), output.getBufferLength()));
+}
+
+TEST(TcrMessage, readSecureObjectPart2) {
+    auto cache = CacheFactory().create();
+    TcrMessageDummy message;
+    uint8_t buffer[] = {0, 0, 0, 3, //Size
+                        0,         //isObj
+                        'a', 'b', 'c' };
+
+    auto input = cache.createDataInput(buffer, sizeof(buffer));
+
+    message.readSecureObjectPart(input, true);
+
+    const auto&& cacheableString = std::dynamic_pointer_cast<CacheableString>(message.getValueRef());
+    ASSERT_THAT(cacheableString.get(), ::testing::NotNull());
+    EXPECT_EQ(cacheableString->toString(), "abc");
+}
+
+TEST(TcrMessage, readSecureObjectPart3) {
+    auto cache = CacheFactory().create();
+    TcrMessageDummy message;
+    uint8_t buffer[] = {0, 0, 0, 0, //Size
+                        0,         //isObj
+                        0 }; //extra data
+
+    auto input = cache.createDataInput(buffer, sizeof(buffer));
+
+    EXPECT_THROW(message.readSecureObjectPart(input), Exception);
+}
+
+TEST(TcrMEssage, getConnectionId) {
+    EXPECT_EQ(TcrMessageDummy().getConnectionId(nullptr), 0);
+}
+
+TEST(TcrMessage, getUniqueId) {
+    EXPECT_EQ(TcrMessageDummy().getUniqueId(nullptr), 0);
+}
+
+TEST(TcrMessage, readExceptionPart) {
+    auto cache = CacheFactory().create();
+    TcrMessageDummy message;
+    auto input = cache.createDataInput(nullptr, 0);
+
+    EXPECT_FALSE(message.readExceptionPart(input, 0, false));
+}
+
+TEST(TcrMessage, DISABLED_writeBytesOnly) {
```

Comment:
Please add a comment about why this test is disabled. Otherwise remove it.

Reply:
A disabled test is a FAILED test!

Reply:
There is what appears to be incorrect behavior when writing the cacheable string to the message. At some point after the object is written, the write pointer is moved back an incorrect number of bytes. The buffer contains a 1 byte DSCode and then a 2 byte size field; the write pointer is moved back to the size field, where the first byte of this 2 byte field is subsequently read as the DSCode. It warrants further investigation.

Reply:
Do we not have any old integration tests that pass through this same code? See coverage reports.

Reply:
This test exists because the coverage reports indicate this code path was not exercised.